pdg/src/info.rs

```diff
@@ -0,0 +1,515 @@
+                greatest_index = node_id_to_check.index();
+                greatest_node_idx = node_id_to_check;
+            }
+            add_children_to_vec(g, &desc_seen, &mut to_view);
+        }
+    }
+    greatest_node_idx
+}
+
+/// Finds the highest-up ancestor of the given node n in g which is reached through Copy, Field, and
+/// Offset, and returns its index as well as the Fields through which it is reached, in order
+/// (the final element is the Field closest to the returned idx)
+fn calc_lineage(g: &Graph, n: &NodeId) -> (NodeId, Vec<Field>) {
+    let mut lineage = Vec::new();
+    let mut n_idx = *n;
+    loop {
+        let node = g.nodes.get(n_idx).unwrap();
+        let parent = match node.source {
+            None => break,
+            Some(p) => p,
+        };
+        n_idx = match node.kind {
+            NodeKind::Offset(_) => parent,
+            NodeKind::Copy => parent,
+            NodeKind::Field(f) => {
+                lineage.push(f);
+                parent
+            }
+            _ => break,
+        };
+    }
+    (n_idx, lineage)
+}
+
+/// Looks for a node which proves that the given node n is not unique. If any is found, it's
+/// immediately returned (no guarantee of which is returned if multiple violate the uniqueness conditions);
+/// otherwise None is returned.
+pub fn check_whether_rules_obeyed(g: &Graph, n: &NodeId) -> Option<NodeId> {
+    let (oldest_ancestor, oldest_lineage) = calc_lineage(g, n);
+    let youngest_descendent = greatest_desc(g, n);
+    let mut to_view = vec![(oldest_ancestor, oldest_lineage)];
+    while let Some((cur_node_id, mut lineage)) = to_view.pop() {
+        if cur_node_id == *n {
+            continue;
+        }
+        if let NodeKind::Field(f) = g.nodes.get(cur_node_id).unwrap().kind {
+            match lineage.pop() {
+                None => continue,
+                Some(top_of_vec) => {
+                    if top_of_vec != f {
+                        continue;
+                    }
+                }
+            }
+        }
+        if lineage.is_empty()
+            && cur_node_id.index() >= n.index()
+            && cur_node_id.index() <= youngest_descendent.index()
+        {
+            return Some(cur_node_id);
+        }
+        to_view.extend(
+            g.nodes
+                .iter_enumerated()
+                .filter_map(|(id, node)| Some((id, node.source?)))
+                .filter(|(_, src_idx)| *src_idx == cur_node_id)
+                .map(|(id, _)| (id, lineage.clone())),
+        );
+    }
+    None
+}
+
+/// Takes a list of graphs, creates a NodeInfo object for each node in each graph, filling it with
+/// the correct flow information.
+pub fn augment_with_info(pdg: &mut Graphs) {
+    let gs: &mut IndexVec<GraphId, Graph> = &mut pdg.graphs;
+    for g in gs {
+        let flow_info = create_flow_info(g);
+        let mut conflicting = determine_non_conflicting(&g,&collect_children(&g),&flow_info);
```

Comment:
`conflicting = determine_non_conflicting` is definitely confusing to me. is one of them misnamed?